### PR TITLE
[Refactor] InternalEngine to always use soft deletes

### PR DESCRIPTION
**Pull request by nknize**
Soft Deletes have been enabled by default since Legacy version 7.0 and made
mandatory in Version 2.0.0. This PR refactors the InternalEngine to always
use soft-deletes. It is a follow on to making soft deletes mandatory in 2.0.0.

### CODE REVIEW

**Comment by nknize** [comment]
Repeat timeout failure reported in #1957 

```
REPRODUCE WITH: ./gradlew ':server:test' --tests "org.opensearch.cluster.routing.MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation" -Dtests.seed=2DE1E5F63CA0EA53 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=el-GR -Dtests.timezone=America/Yakutat -Druntime.java=17
```
```
org.opensearch.cluster.routing.MovePrimaryFirstTests > testClusterGreenAfterPartialRelocation FAILED
    java.lang.AssertionError: timed out waiting for green state
        at __randomizedtesting.SeedInfo.seed([2DE1E5F63CA0EA53:7E2B819317E49547]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureColor(OpenSearchIntegTestCase.java:986)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureGreen(OpenSearchIntegTestCase.java:925)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureGreen(OpenSearchIntegTestCase.java:914)
        at org.opensearch.cluster.routing.MovePrimaryFirstTests.testClusterGreenAfterPartialRelocation(MovePrimaryFirstTests.java:115)
```

**Comment by nknize** [comment]
start gradle check

File: qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java [hunk 1]
```diff
@@ -1344,8 +1350,8 @@ public void testOperationBasedRecovery() throws Exception {
             if (minimumNodeVersion().before(Version.V_2_0_0)) {
                 settings.put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), randomBoolean());
             }
-            createIndex(index, settings.build());
-            ensureGreen(index);
+            final String mappings = randomBoolean() ? "\"_source\": { \"enabled\": false}" : null;
+            createIndex(index, settings.build(), mappings);            ensureGreen(index);
```
Comment (reta) [inline qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java:1354]: Sorry, nit-picking, seems like `ensureGreen(index);` should be moved to the next line

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -1503,6 +1493,18 @@ public DeleteResult delete(Delete delete) throws IOException {
                         plan.currentlyDeleted == false
                     );
                 }
+                if (plan.deleteFromLucene) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1496]: Shouldn't it be moved into `if (plan.deleteFromLucene || plan.addStaleOpToLucene)` block? Not a big deal, but we already have a check for  `plan.deleteFromLucene` in there:

```
if (plan.deleteFromLucene || plan.addStaleOpToLucene) {
    deleteResult = deleteInLucene(delete, plan);
    if (plan.deleteFromLucene) {
     ....
    }
} ...
```

File: qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java (hunk 1 above)
Comment (nknize) [reply qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java:1354]: Good nit-pick! Correcting typo

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (nknize) [reply server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1496]: Curious if the JVM already optimizes for that? Either way, I won't argue w/ saving the extra check :)

Comment (reta) [reply server/src/main/java/org/opensearch/index/engine/InternalEngine.java]: It is more from readability perspective, the things which happen when `plan.deleteFromLucene ` are kept closer to each other, otherwise it is difficult to build a picture (at least when I read the code), thank you!

**Review by reta — APPROVED** [review]

File: qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java [hunk 3]
```diff
@@ -1344,7 +1350,8 @@ public void testOperationBasedRecovery() throws Exception {
             if (minimumNodeVersion().before(Version.V_2_0_0)) {
                 settings.put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), randomBoolean());
             }
-            createIndex(index, settings.build());
+            final String mappings = randomBoolean() ? "\"_source\": { \"enabled\": false}" : null;
```
Comment (andrross) [inline qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java:1353]: How is this change related to soft deletes?

**Review by andrross — APPROVED** [review]

File: qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java (hunk 3 above)
Comment (nknize) [reply qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/FullClusterRestartIT.java]: It's not related to this PR. While not detrimental this was cruft left over from me trying something unrelated. Thanks for catching. I've removed.

**Comment by nknize** [comment]
Another #1908 failure

```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.gateway.QuorumGatewayIT.testQuorumRecovery" -Dtests.seed=DBB99468CB97443F -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-Latn-ME -Dtests.timezone=Atlantic/Stanley -Druntime.java=17
```

```
org.opensearch.gateway.QuorumGatewayIT > testQuorumRecovery FAILED
    java.lang.AssertionError: timed out waiting for green state
        at __randomizedtesting.SeedInfo.seed([DBB99468CB97443F:3E86A8BB6AEA9324]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureColor(OpenSearchIntegTestCase.java:986)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureGreen(OpenSearchIntegTestCase.java:925)
        at org.opensearch.test.OpenSearchIntegTestCase.ensureGreen(OpenSearchIntegTestCase.java:914)
        at org.opensearch.gateway.QuorumGatewayIT.testQuorumRecovery(QuorumGatewayIT.java:111)
```

**Comment by nknize** [comment]
start gradle check
